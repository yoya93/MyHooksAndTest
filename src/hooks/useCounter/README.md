# useCounter Hook

Ejemplo de uso:

```
    const { state, incrementar, decrement, reset } = useCounter(10);
```

useCounter() // recibe un valor por defecto
