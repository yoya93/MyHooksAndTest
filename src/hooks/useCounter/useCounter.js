import { useState } from "react";

export const useCounter = (initialState = 10) => {
  const [state, setstate] = useState(initialState);

  const incrementar = (factor) => {
    setstate(state + factor);
  };
  const decrementar = (factor) => {
    setstate(state - factor);
  };

  const reset = () => setstate(initialState);
  return {
    state,
    incrementar,
    decrementar,
    reset,
  };
};
