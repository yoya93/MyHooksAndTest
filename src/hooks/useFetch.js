import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {
  const isMounted = useRef(true);
  const [apiData, setapiData] = useState({
    data: null,
    loading: true,
    error: null,
  });

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setapiData({
      loading: true,
      data: null,
      error: null,
    });
    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        isMounted.current &&
          setapiData({
            loading: false,
            data,
            error: null,
          });
      })
      .catch(() => {
        setapiData({
          data: null,
          loading: false,
          error: "No se pudo cargar la info",
        });
      });
  }, [url]);

  return apiData;
};
