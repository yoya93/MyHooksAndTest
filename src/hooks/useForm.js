// este hooks recibe un objeto con los key de los input que
//quiera usted seguir en su formulario y retorna el valor actual de este
//y una funcion para que pueda ser modificado este state desde el padre
// una vez recibido el  value en el component padre debe destructurarlo

import { useState } from "react";

export const useForm = (initialState = {}) => {
  const [value, setValue] = useState(initialState);

  const reset = () => {
    setValue(initialState);
  };

  const handleInputChange = ({ target }) => {
    setValue({
      ...value,
      [target.name]: target.value,
    });
  };

  return [value, handleInputChange, reset];
};
