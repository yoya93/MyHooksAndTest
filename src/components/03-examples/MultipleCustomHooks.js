import React from "react";
import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";

import "../02-useEffect/simpleForm.css";
export const MultipleCustomHooks = () => {
  const initial = 1;
  const { state, incrementar, decrementar } = useCounter(initial);

  const endpoint = `https://www.breakingbadapi.com/api/quotes/${state}`;

  const { loading, data } = useFetch(endpoint);
  const { author, quote } = !!data && data[0];

  return (
    <>
      <h1>BreakingBad Quotes</h1>
      <hr />

      {loading ? (
        <div className="alert alert-info text-center">Loading</div>
      ) : (
        <blockquote className="blockqoute text-right">
          <p className="mb-0"> {quote}</p>
          <footer className="blockquote-footer">{author}</footer>
        </blockquote>
      )}

      <button
        disabled={state === 1 ? true : false}
        onClick={() => decrementar(initial)}
        className="btn btn-primary"
      >
        Anterior
      </button>
      <button onClick={() => incrementar(initial)} className="btn btn-primary ">
        Next
      </button>
    </>
  );
};
