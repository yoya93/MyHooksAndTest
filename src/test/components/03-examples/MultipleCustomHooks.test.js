import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";

jest.mock("../../../hooks/useCounter");

jest.mock("../../../hooks/useFetch"); // no me importa la implementacion del hooks porke ya esta probado
//lo ke me interesa es utilizar lo ke me retorna este hooks por esa razon la paso por el mock
// de esta forma el useFetch que desconectado de lo ke el realiza normalmente y queda dispuesto a que yo maneje los datos ke devuelve

describe("Pruebas en el <MultipleCustomHooks>", () => {
  useCounter.mockReturnValue({
    state: 1,
    incrementar: (n) => {},
  });

  test("debe hacer match el componente MultipleCustomHooks ", () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot();
  });

  test(" should debe mostrar la informacion  ", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "fernando",
          quote: "hola mundo",
        },
      ],
      loading: false,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find(".alert").exists()).toBe(false);
    expect(wrapper.find(".mb-0").text().trim()).toBe("hola mundo");
    expect(wrapper.find("footer").text().trim()).toBe("fernando");
  });
});
