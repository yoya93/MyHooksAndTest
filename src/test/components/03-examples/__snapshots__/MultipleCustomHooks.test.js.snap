// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <MultipleCustomHooks> debe hacer match el componente MultipleCustomHooks  1`] = `
<Fragment>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading
  </div>
  <button
    className="btn btn-primary"
    disabled={true}
    onClick={[Function]}
  >
    Anterior
  </button>
  <button
    className="btn btn-primary "
    onClick={[Function]}
  >
    Next
  </button>
</Fragment>
`;
